src/Graph.cpp

```diff
@@ -1166,6 +1152,14 @@ unordered_map<uint,string> Graph::getIndexToNodeNameMap() const {
     return res;
 }
 
+bool Graph::hasSelfLoop(uint source) const {
+    vector<uint> neighbors = adjLists[source];
+    if(std::find(neighbors.begin(), neighbors.end(), source) != neighbors.end()) {
```

Comment:
This seems terribly expensive. Why not just check the single bit that's along the diagonal of the adjacency matrix?

Note that the method you have here *will* be necessary if SPARSE is true; but if SPARSE is false then just perform the O(1) operation of looking at the diagonal of the adjacency matrix.